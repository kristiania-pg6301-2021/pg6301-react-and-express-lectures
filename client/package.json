{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "parcel src/index.html",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-router": "^5.2.0",
    "react-router-dom": "^5.2.0"
  },
  "devDependencies": {
    "parcel": "^2.0.0-beta.2",
    "prettier": "^2.2.1"
  }
}
