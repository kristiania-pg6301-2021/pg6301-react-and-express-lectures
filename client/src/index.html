<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>KristianiaQuiz</title>
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
