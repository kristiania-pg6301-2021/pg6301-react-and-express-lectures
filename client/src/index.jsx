import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter, Link } from "react-router-dom";
import { Route, Switch } from "react-router";

function Application() {
  return (
    <BrowserRouter>
      <header>
        <Link to={"/"}>KristianiaQuiz</Link>
      </header>
      <main>
        <Switch>
          <Route path="/login">
            <h1>Login</h1>
          </Route>
          <Route path="/" exact>
            <h1>Home page</h1>
            <ul>
              <li>
                <Link to={"/login"}>Login</Link>
              </li>
            </ul>
          </Route>
        </Switch>
      </main>
    </BrowserRouter>
  );
}

ReactDOM.render(<Application />, document.getElementById("app"));
