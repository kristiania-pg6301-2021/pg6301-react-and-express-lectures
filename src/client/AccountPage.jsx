import React, { useEffect, useState } from "react";

function useLoader(loader, deps) {
  const [loading, setLoading] = useState();
  const [data, setData] = useState();
  const [error, setError] = useState();

  async function reload() {
    setLoading(true);
    setError(undefined);
    setData(undefined);
    try {
      setData(await loader());
    } catch (e) {
      setError(e);
    } finally {
      setLoading(false);
    }
  }

  useEffect(reload, deps || []);

  return { loading, data, error, reload };
}

export function AccountPage() {
  const { loading, data, error } = useLoader(async () => {
    const res = await fetch("/api/account");
    if (!res.ok) {
      throw new Error("Failed to GET " + res.url + ": " + res.statusText);
    }
    return res.json();
  });

  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return <div>Error: {error.toString()}</div>;
  }

  return (
    <>
      <h1>Account</h1>
      <div>You have ${data.account.balance}</div>
    </>
  );
}
