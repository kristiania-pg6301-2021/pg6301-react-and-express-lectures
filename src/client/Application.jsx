import { BrowserRouter, Link } from "react-router-dom";
import { Route, Switch } from "react-router";
import { ProfilePage } from "./ProfilePage";
import * as React from "react";

export function Application() {
  return (
    <BrowserRouter>
      <header>
        <Link to={"/"}>Home</Link>
      </header>
      <Switch>
        <Route path={"/"} exact>
          <h1>Home page</h1>
          <ul>
            <li>
              <Link to={"/profile"}>Profile</Link>
            </li>
            <li>
              <Link to={"/chat"}>Chat</Link>
            </li>
            <li>
              <Link to={"/login"}>Login</Link>
            </li>
          </ul>
        </Route>
        <Route path={"/profile"}>
          <ProfilePage />
        </Route>
        <Route path={"/chat"}>
          <h1>Chat page</h1>
        </Route>
        <Route path={"/login"}>
          <h1>Login page</h1>
        </Route>
      </Switch>
    </BrowserRouter>
  );
}