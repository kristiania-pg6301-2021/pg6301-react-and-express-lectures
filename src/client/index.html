<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Hello Kristiania</title>
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
