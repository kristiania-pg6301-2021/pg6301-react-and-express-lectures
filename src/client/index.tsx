import * as React from "react";
import * as ReactDOM from "react-dom";

import {BrowserRouter} from "react-router-dom";
import {Route, Switch} from "react-router";

interface Quiz {
    question: string;
    alternatives: string[];
}


function HomePage() {
    return <>
        <h1>Welcome to the quiz</h1>
        <button>Start new quiz</button>
    </>
}

function QuizPage({quizFactory}: {quizFactory(): Quiz}) {
    const quiz = quizFactory();
    const {question, alternatives} = quiz;
    return <>
        <h1>Quiz</h1>
        <div>Question: {question}</div>
        {alternatives.map(a =>
            <button>{a}</button>
        )}
    </>
}

function App() {
    const quiz: Quiz = {
        question: "Are you happy?",
        alternatives: ["yes", "no", "maybe"]
    }
    return <BrowserRouter>
        <Switch>
            <Route path="/quiz">
                <QuizPage quizFactory={() => quiz} />
            </Route>
            <Route exact path="/">
                <HomePage />
            </Route>
            <Route>
                Not found
            </Route>
        </Switch>
    </BrowserRouter>;
}


ReactDOM.render(<App />, document.getElementById("app"));