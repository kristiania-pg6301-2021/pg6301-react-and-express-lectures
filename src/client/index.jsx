import React from "react";
import ReactDOM from "react-dom";
import { fetchJSON, postJSON } from "./http";
import { Application } from "./Application";
import { BrowserRouter } from "react-router-dom";

const bookApi = {
  listBooks: async () => await fetchJSON("/api/books"),
  getBook: async (id) => await fetchJSON(`/api/books/${id}`),
  createBook: async ({ title, author, year }) =>
    await postJSON("/api/books", { title, author, year }),
  updateBook: async (id, { title, author, year }) =>
    await postJSON(`/api/books/${id}`, { title, author, year }, "PUT"),
};

ReactDOM.render(
  <BrowserRouter>
    <Application bookApi={bookApi} />
  </BrowserRouter>,
  document.getElementById("root")
);
