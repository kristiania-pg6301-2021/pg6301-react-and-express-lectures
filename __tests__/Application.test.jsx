import { Application } from "../src/client/Application";
import * as ReactDOM from "react-dom";
import * as React from "react";
import { act } from "react-dom/test-utils";
import { MemoryRouter } from "react-router-dom";

describe("application", () => {
  it("can show home page", async () => {
    const container = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <Application />
      </MemoryRouter>,
      container
    );
    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("h1").textContent).toEqual(
      "Book application"
    );
  });

  it("can navigate to create book page", async () => {
    const container = document.createElement("div");
    await act(async () => {
      await ReactDOM.render(
        <MemoryRouter>
          <Application />
        </MemoryRouter>,
        container
      );
    });
    const createBookLink = [...container.querySelectorAll("a")].find(
      (a) => a.textContent === "Add a book"
    );
    await act(async () => {
      await createBookLink.dispatchEvent(
        new MouseEvent("click", { bubbles: true })
      );
    });
    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("h1").textContent).toEqual(
      "Create new book"
    );
  });
});
