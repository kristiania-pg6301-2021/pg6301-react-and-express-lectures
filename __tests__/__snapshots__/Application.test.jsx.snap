// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application can navigate to create book page 1`] = `"<nav><a href=\\"/\\">Home</a></nav><main><form><h1>Create new book</h1><div><label>Title: <input type=\\"text\\" value=\\"\\"></label></div><div><label>Author: <input type=\\"text\\" value=\\"\\"></label></div><div><label>Year: <input type=\\"number\\" value=\\"\\"></label></div><button>Submit</button></form></main>"`;

exports[`application can show home page 1`] = `"<nav><a href=\\"/\\">Home</a></nav><main><h1>Book application</h1><ul><li><a href=\\"/books\\">List books</a></li><li><a href=\\"/create\\">Add a book</a></li></ul></main>"`;
